### DISCUSSION TITLE
Update on the rust port

### DISCUSSION THREAD

**Discussion by faho** [fish-shell/fish-shell#10123] [Announcements]
As [asked by](https://github.com/fish-shell/fish-shell/pull/9512#issuecomment-1826761540) @PoignardAzur:

> some people over at r/rust are curious how the rewrite is going?

In short: We're "mostly" done. Quantifying that isn't all that easy because there are separate measurements - do you use lines or bytes, do you use rust vs C++ or do you use the amount of C++ that was *removed*?

Depending on how you count, we're at anywhere between 60% (c++ sloc removed) and 74% (rust sloc vs c++ sloc) done.

*And* a lot of that is scaffolding we've duplicated that can be removed once the last caller is gone, so you get a bit of a domino effect. E.g. we have already translated wgetopt, but kept the C++ version around until the last caller is translated, which is *simpler* than adding translation gunk to all the C++ callers.

What is left to do?

The last big component that remains to be translated is the reader with the input system. This is 3.5k for the reader plus 1.3k for the screen handling plus 1k line for input plus 700 lines for the pager. This is pretty tightly coupled.

After that the rest is some stragglers and the de-C++ifying of the build system - removing cmake, getting cargo to do everything we want (cmake still substantially drives the build, and we need e.g. cargo to build and install the documentation and completion files. this isn't just compiling a bunch of rust, we have a lot of auxilliary stuff!)

This isn't *really* the sort of thing drive-by contributions are good for, I'm afraid.

--------------

Alright, that's the update, now for answering some questions / correcting some misconceptions I've seen in what I assume is [this reddit thread](https://old.reddit.com/r/rust/comments/183odpt/media_fishshell_has_finally_been_converted_into_a/):

1. We will be removing *all* of the C++.
2. A windows port is not planned, and has never been planned. C++ is also perfectly well-supported on Windows, so rust doesn't really help here. The language has never been the problem with a windows port, the problem is that windows is not unix!
3. We won't be changing the name or the mascot. We probably won't add "written in rust" to the tag line (but we might mention it on https://fishshell.com/)
4. We have no real idea what the performance of the finished port is going to be. Early results are encouraging, but it is entirely possible this will be ~20% slower in some cases. Even once the initial port is done, the project won't be. There's a loooot to do.

**Comment by PoignardAzur** [comment]
> We won't be changing the name or the mascot.

But but but turbofish!

**Reply by itslychee** [reply]
gotta go fast :dash:

**Reply by Xiretza** [reply]
Honestly that wouldn't be a terrible codename for the initial Rust release, if fish did those ;)

**Reply by shtrophic** [reply]
> but it is entirely possible this will be ~20% slower in some cases

"turbo"fish

**Comment by faho** [comment]
Another thing:

We plan on not doing any partial-rust release.

That means we would be doing e.g. fish 4.0 as fully rust and zero C++, and I think, contrary to what we usually do that warrants a beta. (Ordinarily we've stopped doing betas and release candidates because they simply don't get any testing).

We also still want to do a 3.7.0 in-between release that is still purely C++, so we have a better jumping off point for platforms that can't handle the rust version. It would be 3.6.1 with some neat changes backported.

**Reply by faho** [reply]
3.7.0 is built in the [Integration_3.7.0 branch](https://github.com/fish-shell/fish-shell/tree/Integration_3.7.0) in case you want to see what's happening there - including the (preliminary) [CHANGELOG](https://github.com/fish-shell/fish-shell/blob/Integration_3.7.0/CHANGELOG.rst).

Highlights:

- A timeout for key combinations, useful for binding "kj" or "jj" to switching mode in vi-mode
- Usability improvements with the history-pager (ctrl-r) - e.g. it'll start with the existing search if you decide to switch from up-arrow to ctrl-r
- Some rendering glitches are fixed
- Faster globs

**Comment by Berrysoft** [comment]
You have done a really hard work to port the code from C++ to Rust. As a user of fish, I admire your perseverance.

I'm also a Windows user. I use fish with MSYS2 environment, which is a fork of cygwin. Fish ran really well on it because it provides POSIX utilities. However, Rust doesn't provide a cygwin or MSYS2 target, which means fish cannot be compiled target cygwin or MSYS2 after rewritten by Rust. The language *was* not a problem with a Windows port, but it *will be*.

I'm just a user, so this comment is just a reminder, that the rewriting work might block some fish users on Windows to update their favorite shell. I'm not blaming the work or the language. Anyway, Windows is not Unix:)

Thanks again for all of your hard work.

**Reply by faho** [reply]
The major blocking issue there is that [cygwin does not have an llvm maintainer](https://github.com/rust-lang/rust/issues/79854#issuecomment-1765574972).

It appears that cygwin is in need of some love, llvm is the second-most popular open source compiler. It is a big and important package.

Until then, you'd have to use WSL on windows, which has a lot of things going for it from our perspective - it's much closer to linux in behavior, which makes handling it much nicer.

But yeah, this is an argument for making a 3.7.0 release with C++ - so cygwin at least gets up to there.

**Reply by Berrysoft** [reply]
It's rather common to cross-compile on Windows, and fish doesn't need LLVM. So LLVM is not the major blocking issue. It is Rust team itself doesn't have much interest on that target. The crates, `libc`, `cc`, `compiler_builtins` etc., have no knowledge about cygwin target.

I don't think it is your responsibility to consider these compatibility issues because fish shell initially didn't plan for Windows. But *please* don't give advice about using WSL. WSL is Linux, not Windows. It won't be better if many works need to be on Windows.

**Reply by LoganDark** [reply]
> It won't be better if many works need to be on Windows.

WSL will still find `.exe`'s in PATH and execute them. What can you do in Cygwin that you can't do with WSL? Is all you need a shared process namespace, or is it access to Win32 APIs directly, or what?

**Reply by nightkr** [reply]
> However, Rust doesn't provide a cygwin or MSYS2 target

The `x86_64-pc-windows-gnu` should be usable with MSYS2. I don't use it regularly (not a Windows user), but I was able to get it to compile some other stuff in MSYS2 a few days ago.

**Reply by RReverser** [reply]
> The `x86_64-pc-windows-gnu` should be usable with MSYS2. I don't use it regularly (not a Windows user), but I was able to get it to compile some other stuff in MSYS2 a few days ago.

Same, that's more or less the designated target for MSYS2. I've used it to build Rust code with some libraries that were ported from Linux but only to MSYS2 (as in, with bunch of other MSYS2 packages as dependencies).

**Reply by Berrysoft** [reply]
> What can you do in Cygwin that you can't do with WSL?

It's obvious that the IO of a native Win32 executable is faster than WSL, when accessing the same Windows drive:) In particular, I don't want the Linux `git` shadow the Windows one because there are other corner cases(encoding, line ending, symlink handling and so on).

> The `x86_64-pc-windows-gnu` should be usable with MSYS2.

That's MinGW not MSYS2. It doesn't provide a `fork` method. However, fish requires it: https://github.com/fish-shell/fish-shell/blob/master/fish-rust/src/fork_exec/postfork.rs

**Reply by RReverser** [reply]
MinGW is just one of subsystems of MSYS2. You can pick the one that works for your needs.

As for fork, it would be implemented by manually creating a new process anyway if emulated by such environment, since Windows doesn't support fork natively, so you might as well contribute such fallback directly to fish.

**Reply by Berrysoft** [reply]
That's right, but fish assumes POSIX strongly. It won't be easy to replace all such stuffs. The easiest solution, I think, is to create a cygwin target for Rust.

**Reply by faho** [reply]
>Windows doesn't support fork natively, so you might as well contribute such fallback directly to fish.

Just so nobody ends up wasting their time: That's the sort of thing where you should ask if it's acceptable before implementing it.

I think we probably wouldn't take it - we don't have a native windows port in part because it would require a lot of stuff like this.

---

Your best chance here is to get a cygwin target for rust, which requires someone to take it up, which requires, from my understanding, someone to maintain llvm for cygwin.

**Reply by Berrysoft** [reply]
I aggree that the best chance is to add a cygwin target, and I also think that llvm for cygwin is not that necessary:) Anyway we can cross compile.

**Reply by RReverser** [reply]
> I think we probably wouldn't take it - we don't have a native windows port in part because it would require a lot of stuff like this.

That's somewhat unfortunate, one of the strong benefits of Rust is that it provides much easier cross-platform compatibility than with C/C++ so this port could bring fish much closer to Windows.

Although that only applies if you use native Rust stdlib & crates. I see fish calls out to libc functions directly, which makes sense for easier porting, but does mean giving up that compatibility for now as the changes required would likely be pretty large as you mentioned. Oh well, maybe someday later down the line, when the codebase uses more Rust APIs.

**Reply by faho** [reply]
Here's hopefully the last thing I'm going to write about a native windows port in some time:

fish is not just a program that happens to run in a unix terminal. It is a unix *shell*, and unix shells are a surprisingly direct translation of the unix API to the screen. To the point that there are some unix APIs I suspect have been added just for the shell.

We don't just use libc directly because it was easier to port, we use libc directly because those are the semantics we need.

Off the top of my head, here's the ways in which fish explicitly touches the unix API:

- terminal handling - ansi escape sequences, but also terminal ownership and modes (tcgetattr/tcsetattr)
- process management - pids, pgroups, waiting, return status, signals
- file permissions - the ugo rwx model with suid and sgid
- users - uid, home directories

Some of these are basically directly exposed in the scripting language - `jobs` lists a process group for every background process, `test` and `path filter` allow checking the ugo rwx model for file permissions, `path mtime` prints the mtime in unix time, `$status` is defined to be an integer from 0 to 255, signal handling expects signals like SIGTERM and SIGHUP. `$EUID` exposes the user id, `~foo` lets you open user's home directories. `ulimit` is a "turn some unix knobs for me" command. `set_color` prints ansi escapes. `umask` is directly coupled with unix rwx permissions.

Others are internal to fish - we do a bunch of cursor movement and escapes to make syntax highlighting and autosuggestions work. We touch the terminal modes after every command both to allow that and to avoid breaking the terminal if a command crashes.

You may note that [we use](https://github.com/fish-shell/fish-shell/blob/9b54e243b18302387cd744147262b86789cef25f/fish-rust/src/builtins/path.rs#L4) the [std::unix::fs::MetadataExt](https://doc.rust-lang.org/stable/std/os/unix/fs/trait.MetadataExt.html) trait to get file modes or users (including uid!), which is unavailable on non-unix platforms. Windows has [one bit in the permissions](https://doc.rust-lang.org/stable/std/fs/struct.Permissions.html) - "readonly".

Fish's basic scripts [expect a general unix userland](https://github.com/fish-shell/fish-shell#dependencies) - tools like `cat` or `rm` or `mkfifo`.

Would it be *possible* to make fish "run" on windows? Probably. Would it be possible while keeping existing scripts working, without changing the core semantics? Probably not.

But the biggest question is: How much would have to be done to allow that? And the answer is: Probably more than you think.

And we do not want to take up that maintenance burden. That's not a matter of waiting for a "down the line, when the codebase uses more Rust APIs", that's a fact of fish being specifically a unix shell, designed for unix.

It is like I said: "The language has never been the problem with a windows port, the problem is that windows is not unix!". C++ works on windows, we could have ported it to windows with C++. Rust might offer nicer abstractions, but fish simply isn't that abstracted from unix.

----

Alright, I *hope* that's that topic covered.

I'm not saying you can't make a shell that works on windows and unix, I'm saying fish is not that shell. To make it into that would require a big rewrite and redesign, and we're not going to do that.

**Reply by BatmanAoD** [reply]
@Berrysoft I'm not sure why you don't think LLVM is required; LLVM is the only stable backend for the Rust compiler.

**Reply by Berrysoft** [reply]
Because if we only want to compile fish to cygwin, an easier way is to cross compile with an *-msvc rustc or *-gnu rustc. They don't require an LLVM on cygwin.

**Reply by RReverser** [reply]
Rust itself still ships with LLVM in either scenario. Those targets only change which standard libraries your binary is linked against - the native Windows ones or the GNU ones provided by MSYS2.

**Reply by BatmanAoD** [reply]
It also changes binary codegen, but yes, LLVM is currently required for all stable toolchains.

**Reply by Berrysoft** [reply]
Yes, and LLVM always support targeting cygwin. Rustc cannot be compiled targeting cygwin because the LLVM version cygwin distributed is a little old. An *-msvc rustc is enough to create a cygwin exe. If the std and other rust crates fixed, I believe we can compile fish targeting cygwin.

**Reply by trajano** [reply]
I don't think we really need a full port of Fish for Windows as there's several POSIXy things that do not have Windows equivalents.  

> Would it be possible while keeping existing scripts working, without changing the core semantics? **definitely not, I don't think there's a logical path to even make this have the word probable**

However, we can probably get a way with partial implementations or totally skip.  I was thinking a fork that would just drop all the extra functionality that fish would require for users that want a __friendlier__ shell:

* namely better command line history (bash's management is dated)
* tab completions 
* richer prompt handling (via oh-my-posh)

Some things I can think that can be done away with for 

* file permissions - for Windows it would be checking if the file starts with `#!` or if the extension in context ends with .cmd, .bat, .exe, .com .  I don't think Windows has a POSIXy way of determining executable state.

    The rest of the permissions I am presuming we can simply do away with as there's no real meaning on Windows.  If the shell fails to render it then it fails.

* users - The UID in Windows is a UUID string, technically if it is really needed it needs to be parsed to a `long` integer, but even that may not work I am presuming the limit of UID in POSIX is just a `short` or `int32`.  

* terminal handling - ANSI escape sequences I think are mostly handled by Windows Terminal now.

* terminal ownership and modes (tcgetattr/tcsetattr) -- maybe can be dropped?

* process management - these would be more if we're running a fish script not regular use.  In which case we still have the `#!/bin/bash` fallback 

Anyway it's just a thought, I just noticed the topic of this convo is for the *rust port*

**Comment by krobelus** [comment]
These are the remaining files:

```
src/fish_indent_common.cpp
src/fish_indent.cpp
src/fish_key_reader.cpp
src/builtins/bind.cpp
src/builtins/commandline.cpp
src/fish_tests.cpp
src/pager.cpp
src/reader.cpp
```

**Comment by adaszko** [comment]
Sorry if this is a tiny bit off-topic.  I'm the author [complgen](https://github.com/adaszko/complgen) -- a tool that (hopefully) makes writing shell completions easier, faster and less error-prone.  It shines especially if you're writing a completion script for a command with subcommands, like `git(1)` (`git fe<TAB>`, `git pu<TAB>`) where it frees you from maintaining a state machine by hand in shell script code.  The tool is based on shell word-based regular grammars which I find to be adequate for at least 99% of real-life shell completions.   The remaining 1% is for tools like `find(1)` whose grammar is non-regular due to its parenthesis use.  Even in those cases `complgen` is usable and useful, just not 100% *accurate*.  My ultimate goal for the best UX is if all shells included *something* like complgen out of the box, so I want to use this occasion of rewriting `fish` in Rust to probe your thoughts on this.  I realise a rewrite is not meant to add new features.  I'd be happy to clear up any aspects of how complgen works.

**Reply by faho** [reply]
> Sorry if this is a tiny bit off-topic

It is off-topic, yes. You're basically asking for an unrelated enhancement.

-----

> git fe<TAB>, git pu<TAB>

That step isn't *really* the one we have a problem with.

It's the *next* - what do you complete for `git add <TAB>`? And the answer there is complex, and basically requires code, and that code will be different between fish and zsh.

----

Anyway, none of this has to do with what programming language fish uses. I'm going to say that I don't think we are interested in adding a second way to write completions at this time.

**Reply by adaszko** [reply]
>> git fe, git pu
> That step isn't really the one we have a problem with.
> It's the next - what do you complete for git add <TAB>? And the answer there is complex, and basically requires code, and that code will be different between fish and zsh.

I may have given an unrepresentative example above.  `git add <TAB>` is *exactly* the type of cases `complgen` solves.  As I mentioned, there's a state machine being maintained automatically.

**Comment by Jeklah** [comment]
May I suggest you do add "written in Rust", purely for giving Rust publicity.
Let it be known the language is growing!

**Reply by faho** [reply]
Rust has 96000 stars on github. It's the [most admired programming language on the StackOverflow developer survey](https://survey.stackoverflow.co/2024/technology#admired-and-desired) for as far as I can tell the 9th year in a row. It's mentioned in [reports by the White House](https://www.whitehouse.gov/oncd/briefing-room/2024/02/26/press-release-technical-report/).

It's doing just fine on publicity.

**Reply by Jeklah** [reply]
Last I checked this isn't highschool and popularity contests aren't the be all and end all of things. We should act accordingly, and not just reference popularity votes, but rather go on how many jobs are available for a language. It's all well and good a language being loved, but it's useless if it's not widely used.

Regarding stars on github...96000 is pretty small for an entire language. A few C++ projects have many, many more stars than the entire Rust language.
[C++ Ranking](https://github.com/EvanLi/Github-Ranking/blob/master/Top100/CPP.md)

As for the white house comment....not everything mentioned in the white house is true, or come to pass...don't believe everything you see on the news etc. So I don't really count that as a valid comment.

**Reply by Berrysoft** [reply]
You are looking for a nonexistent problem. The project has been written in what the maintainers like. If you don't like Rust, you can maintain the old C++ one. That's how open-source projects work.

**Reply by Jeklah** [reply]
What makes you think I don't like Rust? I was asking it gets more publicity, in the hope more jobs appear!

I was just pointing out that relying on popularity contests and what the white house says aren't very good/reliable.

**Reply by Berrysoft** [reply]
Then we agree that Rust is a good language for fish developers. I think it's enough:)

As for the popularity of Rust, I think that's the duty of Rust Foundation.

**Reply by Jeklah** [reply]
Yes we agree on that, I was never saying it wasn't a good language for fish developers :). Personally I think it will take over C++ for everything in the long (long long) term, as C++ took over Haskell and Fortran...albeit even some systems still use those...

Yes it possibly is one of the duties of the Rust Foundation, but we can all help in addition :)

**Reply by faho** [reply]
Alright, I'm going to have to ask you to please stop or take it elsewhere.

The point is we're not adding "written in rust" to the tagline and that's it.
